README.md
FetchParse.js
=============
[![NPM version][npm-badge]](https://www.npmjs.com/package/fetch-parse)
[![Build status][travis-badge]](https://travis-ci.org/moll/js-fetch-parse)

FetchParse.js is a mixin for the [Fetch API][fetch] for browsers and Node.js that **automatically buffers response bodies** and assigns them as `body` on the response object (saving you from extra `res.text()` or `res.json()` calls). It can also **optionally parse response bodies** based on the `Content-Type` header (using [MediumType.js][medium-type] for media types).

[npm-badge]: https://img.shields.io/npm/v/fetch-parse.svg
[travis-badge]: https://travis-ci.org/moll/js-fetch-parse.png?branch=master
[fetch]: https://developer.mozilla.org/en/docs/Web/API/Fetch_API
[medium-type]: https://github.com/moll/js-medium-type


Installing
----------
```sh
npm install fetch-parse
```

FetchParse.js follows [semantic versioning](http://semver.org), so feel free to depend on its major version with something like `>= 1.0.0 < 2` (a.k.a `^1.0.0`).


Using
-----
Pass the native `fetch` function to `fetchParse` to get back a new `fetch` function that will buffer all responses and assign the buffered response body to the `body` property on the response object. It'll also parse JSON responses (those whose `Content-Type` matches `application/json` or `*/*+json`):

```javascript
var fetchParse = require("fetch-parse")
var apiFetch = fetchParse(fetch)
```

The above is equal to calling `fetchParse` with the wildcard media type:

```javascript
var fetchParse = require("fetch-parse")
var apiFetch = fetchParse(fetch, {"*/*": true})
```

Now, calling `apiFetch` will give you back a `Promise` with a regular `Response` object, but with the `body` property assigned to the buffered response body:

```javascript
apiFetch("/users").then(function(res) {
  // If Content-Type of /users was application/json and it returned
  // an array, you could start using it right there:
  res.body[0].email
})
```

For built-in parsers that FetchParse.json includes, see below.

### Parsers
To configure what media types FetchParse.js buffers and parses, pass an object of media type patterns and `true`s. The `true` value there indicates you'd like to use the default parser. If you need a custom parser, see [Custom Parsers](#custom-parsers).

```javascript
var fetchParse = require("fetch-parse")

var apiFetch = fetchParse(fetch, {
  "text/markdown": true,
  "json": true,
})
```

FetchParse.js uses [MediumType.js][medium-type]'s [`MediumType.prototype.match`](https://github.com/moll/js-medium-type/blob/master/doc/API.md#MediumType.prototype.match) to match the parser pattern against the response's `Content-Type` header. The buffering and parsing behavior is dependent on `Content-Type` because that allows enabling default parsing of every response with `{"*/*": true}`, but still get `text/plain` parsed as a string, `application/json` as a JavaScript object etc.

#### Default Parsers

`Content-Type`     | Description
-------------------|------------
`text/*`           | Calls `res.text()` and returns a string.
`application/json` | Equivalent to `res.json()` and returns a parsed object.
`*/*+json`         | _See `application/json` above._
`application/xml`  | Calls `res.text()` and returns a string.
`*/*+xml`          | _See `application/xml` above._
`*/*`              | Calls `res.arrayBuffer()` and returns an `ArrayBuffer`.

#### Shorthands
Aside from regular media type patterns, FetchParse.js has two shorthands for JSON and XML. That's because [RFC 6839][rfc6839] specifies standardized suffices for those two in addition to `application/json` and `application/xml`. Using `json` for example saves you from having to specify both patterns explicitly.

Shorthand | Types
----------|------
`json`    | `application/json`, `*/*+json`
`xml`     | `application/xml`, `*/*+xml`

[rfc6839]: https://tools.ietf.org/html/rfc6839

### Custom Parsers
To use a custom parser for some media types, pass an object mapping a media type pattern to that parse function. The parse function must take a `Fetch.Request` object and return a `Promise` of a body. FetchParse.js will itself assign that to the `body` property on the response object.

```javascript
var fetchParse = require("fetch-parse")

var apiFetch = fetchParse(fetch, {
  "text/html": parseHtml,
  "image/*": parseImage,
  "application/vnd.example.model+json": parseModel
})
```

You can implement `parseHtml` by first calling `Response.prototype.text` to get the plain-text version and then creating a `DocumentFragment` from it:

```javascript
function parseHtml(res) {
  return res.text().then(function(html) {
    return document.createRange().createContextualFragment(html)
  })
}
```

For converting bytes of an image to a URL you could later use with `<img src="" />`, you could call `Response.prototype.arrayBuffer` and create a Blob URL from it (e.g. `blob:d3958f5c-0777-0845-9dcf-2cb28783acaf`):

```javascript
function parseImage(res) {
  return res.arrayBuffer().then(function(body) {
    var blob = new Blob([body], {type: res.headers.get("content-type")})
    return URL.createObjectURL(blob)
  })
}
```

To parse JSON into a your own model instance, you can first use `Response.prototype.json` to have it parsed into an object and then instantiate your model with it:

```javascript
function Model(attrs) { this.attributes = attrs }

function parseModel(res) {
  return res.json().then(function(attrs) { return new Model(attrs) })
}
```

### Browser
Browsers have the Fetch API available at `window.fetch`:

```javascript
var fetchParse = require("fetch-parse")
var apiFetch = fetchParse(window.fetch)
```

### Node.js
Node.js doesn't have a built-in implementation of the Fetch API, but you can use any library with a compatible interface, such as my [Fetch/Off.js][fetch-off] or [node-fetch][node-fetch]:

[fetch-off]: https://github.com/moll/node-fetch-off
[node-fetch]: https://github.com/bitinn/node-fetch

```javascript
var fetch = require("fetch-off")
var fetchParse = require("fetch-parse")
var apiFetch = fetchParse(fetch, "https://example.com")
```


License
-------
FetchParse.js is released under a *Lesser GNU Affero General Public License*, which in summary means:

- You **can** use this program for **no cost**.
- You **can** use this program for **both personal and commercial reasons**.
- You **do not have to share your own program's code** which uses this program.
- You **have to share modifications** (e.g. bug-fixes) you've made to this program.

For more convoluted language, see the `LICENSE` file.


About
-----
**[Andri Möll][moll]** typed this and the code.  
[Monday Calendar][monday] supported the engineering work.

If you find FetchParse.js needs improving, please don't hesitate to type to me now at [andri@dot.ee][email] or [create an issue online][issues].

[email]: mailto:andri@dot.ee
[issues]: https://github.com/moll/js-fetch-parse/issues
[moll]: http://themoll.com
[monday]: https://mondayapp.com


CHANGELOG.md
## 1.1.0 (Nov 20, 2016)
- Adds `text/xml` to the list of XML content-types.

## 1.0.0 (Nov 16, 2016)
- Renames to FetchParse.js.
- Adds support for passing in an object of types and parsers.  
  For a parsers pass either a function taking a `Response` object and returning a `Promise` with the body or `true` to use the default parser for that type.
- Removes support for passing in an array of types to use default parsers.
- Adds XML support.

## 0.3.0 (Jul 31, 2016)
- Sets `body` to what `arrayBuffer` returns when type not text or JSON.

## 0.2.1 (Jan 2, 2016)
- Assigns the `body` property with `Object.defineProperty` to work on Fetch
  implementations that already have `body` as a getter.

## 0.2.0 (Dec 17, 2015)
- Skips parsing again if body already used (`res.bodyUsed`).  
  This also means `body` _won't be_ set `undefined` in cases where the
  Content-Type is not recognized.
- Allows passing in an array of media types to be read and possibly parsed.  
  Only JSON is actually parsed. The rest are read to `res.body`.
- Reads `text/javascript` as regular text, not JSON.
- Skips parsing JSON if no actual content is returned (such as in response to
  a `HEAD`) and sets `body` to `undefined`.
- Sets the `body` of a response (assigned to `err.response`) to the invalid JSON
  should an error occur.

## 0.1.338 (Dec 11, 2015)
- Skips parsing when response 304 Not Modified.  
  This fixes fetching from servers that send a Content-Type along with 304.
- Skips parsing when response 204 No Content.  
  A fail-safe similar to the above.
- Sets a non-enumerable `response` property containing the request's `Response`
  on the `SyntaxError` should JSON parsing fail.

## 0.1.337 (Dec 11, 2015)
- Scoped release for personal use and profit.


test/fetch.js
var BODY_USED_ERR = "Body has already been consumed."

try {
  var self = global.self = {}
  require("whatwg-fetch")
  exports = module.exports = self.fetch
  exports.Headers = self.Headers
  exports.Request = self.Request
  exports.Response = self.Response
}
finally { delete global.self }

var Response = exports.Response

Response.prototype.arrayBuffer = function() {
  if (this.bodyUsed) Promise.reject(new TypeError(BODY_USED_ERR))
  var array = new Uint8Array(new Buffer(this._bodyText, "binary"))
  return Promise.resolve(array.buffer)
}

// Cannot reload the polyfill as it mutates world state and doesn't return
// exports.
require("require-guard")(module.id)


test/index_test.js
var Sinon = require("sinon")
var Fetch = require("./fetch")
var FetchParse = require("..")
var fetch = FetchParse(Fetch)
var URL = "http://example.com"
var GIF = new Buffer("R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs=", "base64")

describe("FetchParse", function() {
  beforeEach(function() {
    var xhr = global.XMLHttpRequest = Sinon.FakeXMLHttpRequest
    xhr.onCreate = Array.prototype.push.bind(this.requests = [])
  })

  it("must return fetch with Headers, Request and Response", function() {
    fetch.Headers.must.equal(Fetch.Headers)
    fetch.Request.must.equal(Fetch.Request)
    fetch.Response.must.equal(Fetch.Response)
  })

  it("must parse all with default parser given */*", function*() {
    var res, headers
    var fetch = FetchParse(Fetch, {"*/*": true})

    res = fetch(URL)
    headers = {"Content-Type": "text/plain"}
    this.requests.pop().respond(200, headers, "Hello")
    ;(yield res).body.must.equal("Hello")

    res = fetch(URL)
    headers = {"Content-Type": "application/json"}
    this.requests.pop().respond(200, headers, JSON.stringify({key: "value"}))
    ;(yield res).body.must.eql({key: "value"})
  })

  it("must parse */* with given parser", function*() {
    var fetch = FetchParse(Fetch, {
      "*/*": function(res) {
        res.must.be.an.instanceof(Fetch.Response)
        return Promise.resolve("Bye")
      }
    })

    var res = fetch(URL)
    var headers = {"Content-Type": "text/plain"}
    this.requests[0].respond(200, headers, "Hello")
    ;(yield res).body.must.equal("Bye")
  })

  it("must parse one with given parser and */* with default", function*() {
    var res, headers

    var fetch = FetchParse(Fetch, {
      "text/markdown": function(res) { return Promise.resolve("# Bye") },
      "*/*": true
    })

    res = fetch(URL)
    headers = {"Content-Type": "text/markdown"}
    this.requests[0].respond(200, headers, "# Hello")
    yield res.must.then.have.property("body", "# Bye")

    res = fetch(URL)
    headers = {"Content-Type": "application/json"}
    this.requests.pop().respond(200, headers, JSON.stringify({key: "value"}))
    ;(yield res).body.must.eql({key: "value"})
  })

  it("must set body if type given exactly", function*() {
    var fetch = FetchParse(Fetch, {"text/markdown": true})
    var res = fetch("/")
    var headers = {"Content-Type": "text/markdown"}
    this.requests[0].respond(200, headers, "# Hello")
    yield res.must.then.have.property("body", "# Hello")
  })

  it("must set body if type given with wildcard", function*() {
    var fetch = FetchParse(Fetch, {"text/*": true})
    var res = fetch("/")
    var headers = {"Content-Type": "text/markdown"}
    this.requests[0].respond(200, headers, "# Hello")
    yield res.must.then.have.property("body", "# Hello")
  })

  it("must not set body if type not given", function*() {
    var fetch = FetchParse(Fetch, {"text/plain": true})
    var res = fetch("/")
    var headers = {"Content-Type": "text/markdown"}
    this.requests[0].respond(200, headers, "Hello")

    res = yield res
    res.must.not.have.property("body")
    yield res.text().must.then.equal("Hello")
  })

  // Just in case protect against an erroneous 204, too.
  it("must not set body when response 204", function*() {
    var res = fetch(URL)
    var headers = {"Content-Type": "application/json"}
    this.requests[0].respond(204, headers, "")

    res = yield res
    res.must.not.have.property("body")
    yield res.text().must.then.equal("")
  })

  // Facebook's API does that: return a Content-Type header, but no body with
  // its 304 Not Modified
  it("must not set body when response 304", function*() {
    var res = fetch(URL)
    var headers = {"Content-Type": "application/json"}
    this.requests[0].respond(304, headers, "")

    res = yield res
    res.must.not.have.property("body")
    yield res.text().must.then.equal("")
  })

  it("must not set body when Content-Type unrecognized", function*() {
    var fetch = FetchParse(Fetch, {"application/json": true})
    var res = fetch(URL)
    var headers = {"Content-Type": "application/fancy"}
    this.requests[0].respond(200, headers, "Hello")

    res = yield res
    res.headers.get("content-type").must.equal("application/fancy")
    res.must.not.have.property("body")
    yield res.text().must.then.equal("Hello")
  })

  it("must not set body when Content-Type invalid", function*() {
    var res = fetch(URL)
    var headers = {"Content-Type": "application///"}
    this.requests[0].respond(200, headers, "Hello")

    res = yield res
    res.headers.get("content-type").must.equal("application///")
    res.must.not.have.property("body")
    yield res.text().must.then.equal("Hello")
  })

  it("must not set body when Content-Type missing", function*() {
    var res = fetch(URL)
    this.requests[0].respond(200, {}, "Hello")

    res = yield res
    res.must.not.have.property("body")
    yield res.text().must.then.equal("Hello")
  })

  it("must not parse multiple times", function*() {
    var fetch = Fetch
    fetch = FetchParse(fetch)
    fetch = FetchParse(fetch)
    var res = fetch("/")

    var headers = {"Content-Type": "application/json"}
    this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
    res = yield res
    res.body.must.eql({key: "value"})
  })

  // Google Chrome uses the `body` property for streaming. It's implemented
  // as an enumerable getter with no setter:
  it("must set body even if a setter", function*() {
    function fetch(url, opts) {
      return Fetch(url, opts).then(function(res) {
        return Object.defineProperty(res, "body", {
          get: Object, configurable: true, enumerable: true
        })
      })
    }

    var res = FetchParse(fetch)("/")
    var headers = {"Content-Type": "application/json"}
    this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
    res = yield res
    res.body.must.eql({key: "value"})
  })

  describe("when Content-Type is text", function() {
    it("must set body by default when Content-Type is text/plain", function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "text/plain"}
      this.requests[0].respond(200, headers, "Hello")
      ;(yield res).body.must.equal("Hello")
    })

    it("must set body when Content-Type is text/plain", function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "text/plain"}
      this.requests[0].respond(200, headers, "Hello")
      ;(yield res).body.must.equal("Hello")
    })

    it("must set body when Content-Type is text/plain; charset=utf-8",
      function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "text/plain; charset=utf-8"}
      this.requests[0].respond(200, headers, "Hello")
      ;(yield res).body.must.equal("Hello")
    })

    it("must set body when Content-Type is text/javascript", function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "text/javascript"}
      this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
      ;(yield res).body.must.equal(JSON.stringify({key: "value"}))
    })

    it("must set body when Content-Length 0", function*() {
      var res = fetch(URL, {method: "HEAD"})
      var headers = {"Content-Type": "text/plain", "Content-Length": "0"}
      this.requests[0].respond(200, headers, "")
      ;(yield res).body.must.equal("")
    })

    it("must set body when response not OK", function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "text/plain"}
      this.requests[0].respond(401, headers, "Hello")
      ;(yield res).body.must.equal("Hello")
    })

    describe("given text/plain", function() {
      it("must set body when Content-Type is text/plain", function*() {
        var fetch = FetchParse(Fetch, {"text/plain": true})
        var res = fetch("/")
        var headers = {"Content-Type": "text/plain"}
        this.requests[0].respond(200, headers, "Hello")
        ;(yield res).body.must.equal("Hello")
      })
    })
  })

  describe("when Content-Type is JSON", function() {
    it("must parse by default when Content-Type is application/json",
      function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "application/json"}
      this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
      ;(yield res).body.must.eql({key: "value"})
    })

    it("must parse by default when Content-Type is application/vnd.foo+json",
      function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "application/vnd.foo+json"}
      this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
      ;(yield res).body.must.eql({key: "value"})
    })

    it("must parse when response not OK", function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "application/json"}
      this.requests[0].respond(401, headers, JSON.stringify({key: "value"}))

      res = yield res
      res.status.must.equal(401)
      res.body.must.eql({key: "value"})
    })

    it("must not parse JSON if not asked", function*() {
      var fetch = FetchParse(Fetch, {"text/plain": true})
      var res = fetch("/")
      var headers = {"Content-Type": "application/json"}
      this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
      res = yield res
      res.must.not.have.property("body")
      yield res.text().must.then.equal(JSON.stringify({key: "value"}))
    })

    // Some implementations respond to HEAD requests with a Content-Type, but
    // Content-Lenght of 0. This saves throwing a parse error.
    it("must set body to undefined when Content-Length is 0", function*() {
      var res = fetch(URL, {method: "HEAD"})
      var headers = {"Content-Type": "application/json", "Content-Length": "0"}
      this.requests[0].respond(200, headers, "")

      res = yield res
      res.must.have.property("body", undefined)
      res.bodyUsed.must.be.true()
    })

    it("must set body to undefined when Content-Length, but no content",
      function*() {
      var res = fetch(URL, {method: "HEAD"})
      var headers = {"Content-Type": "application/json", "Content-Length": "13"}
      this.requests[0].respond(200, headers, "")

      res = yield res
      res.must.have.property("body", undefined)
      res.bodyUsed.must.be.true()
    })

    // This was a released bug with Remote that I noticed on Nov 25, 2014 and
    // was related to parsing errors not being caught and passed to the
    // response error handler. Nor its related promise.
    it("must reject with SyntaxError given invalid JSON", function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "application/json"}
      this.requests[0].respond(200, headers, "{\"foo\": ")

      var err
      try { yield res } catch (ex) { err = ex }
      err.must.be.an.error(SyntaxError, /Unexpected end/)
      err.must.have.nonenumerable("response")
      err.response.must.be.an.instanceof(Fetch.Response)
      err.response.must.have.property("body", "{\"foo\": ")
    })

    describe("given json", function() {
      var fetch = FetchParse(Fetch, {"json": true})

      it("must parse when Content-Type is application/json",
        function*() {
        var res = fetch(URL)
        var headers = {"Content-Type": "application/json"}
        this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
        ;(yield res).body.must.eql({key: "value"})
      })

      it("must parse with parser when Content-Type is application/json",
        function*() {
        var fetch = FetchParse(Fetch, {
          "json": function(res) {
            return res.json().then(function(obj) { return {lock: obj} })
          }
        })

        var res = fetch(URL)
        var headers = {"Content-Type": "application/json"}
        this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
        ;(yield res).body.must.eql({lock: {key: "value"}})
      })

      it("must parse when Content-Type is application/json; charset=utf-8",
        function*() {
        var res = fetch(URL)
        var headers = {"Content-Type": "application/json; charset=utf-8"}
        this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
        ;(yield res).body.must.eql({key: "value"})
      })

      it("must parse when Content-Type is application/vnd.foo+json",
        function*() {
        var res = fetch(URL)
        var headers = {"Content-Type": "application/vnd.foo+json"}
        this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
        ;(yield res).body.must.eql({key: "value"})
      })

      it("must parse when Content-Type is application/vnd.foo+json; v=1",
        function*() {
        var res = fetch(URL)
        var headers = {"Content-Type": "application/vnd.foo+json; v=1"}
        this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
        ;(yield res).body.must.eql({key: "value"})
      })
    })

    describe("given application/json", function() {
      it("must parse JSON when Content-Type is application/json", function*() {
        var fetch = FetchParse(Fetch, {"application/json": true})

        var res = fetch(URL)
        var headers = {"Content-Type": "application/json"}
        this.requests[0].respond(200, headers, JSON.stringify({key: "value"}))
        ;(yield res).body.must.eql({key: "value"})
      })
    })
  })

  describe("when Content-Type is XML", function() {
    it("must set body by default when Content-Type is application/xml",
      function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "application/xml"}
      this.requests[0].respond(200, headers, "<name>John</name>")
      ;(yield res).body.must.equal("<name>John</name>")
    })

    it("must set body by default when Content-Type is application/vnd.foo+xml",
      function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "application/vnd.foo+xml"}
      this.requests[0].respond(200, headers, "<name>John</name>")
      ;(yield res).body.must.equal("<name>John</name>")
    })

    describe("given xml", function() {
      var fetch = FetchParse(Fetch, {"xml": true})

      it("must set body when Content-Type is application/xml", function*() {
        var res = fetch(URL)
        var headers = {"Content-Type": "application/xml"}
        this.requests[0].respond(200, headers, "<name>John</name>")
        ;(yield res).body.must.equal("<name>John</name>")
      })

      it("must parse given parser", function*() {
        var fetch = FetchParse(Fetch, {
          "xml": function(res) {
            return res.text().then(function(obj) { return {name: "John"} })
          }
        })

        var res = fetch(URL)
        var headers = {"Content-Type": "application/xml"}
        this.requests[0].respond(200, headers, "<name>John</name>")
        ;(yield res).body.must.eql({name: "John"})
      })

      it("must parse when Content-Type is application/xml; charset=utf-8",
        function*() {
        var res = fetch(URL)
        var headers = {"Content-Type": "application/xml; charset=utf-8"}
        this.requests[0].respond(200, headers, "<name>John</name>")
        ;(yield res).body.must.equal("<name>John</name>")
      })

      it("must set body when Content-Type is text/xml", function*() {
        var res = fetch(URL)
        var headers = {"Content-Type": "text/xml"}
        this.requests[0].respond(200, headers, "<name>John</name>")
        ;(yield res).body.must.equal("<name>John</name>")
      })

      it("must parse when Content-Type is application/vnd.foo+xml",
        function*() {
        var res = fetch(URL)
        var headers = {"Content-Type": "application/vnd.foo+xml"}
        this.requests[0].respond(200, headers, "<name>John</name>")
        ;(yield res).body.must.equal("<name>John</name>")
      })
    })

    describe("given application/xml", function() {
      it("must set body if Content-Type is application/xml", function*() {
        var fetch = FetchParse(Fetch, {"application/xml": true})

        var res = fetch(URL)
        var headers = {"Content-Type": "application/xml"}
        this.requests[0].respond(200, headers, "<name>John</name>")
        ;(yield res).body.must.equal("<name>John</name>")
      })
    })
  })

  describe("when Content-Type is binary", function() {
    it("must set body by default when Content-Type is application/octet-stream",
      function*() {
      var res = fetch(URL)
      var headers = {"Content-Type": "application/octet-stream"}
      this.requests[0].respond(200, headers, GIF.toString("binary"))
      res = yield res
      res.body.must.be.an.instanceof(ArrayBuffer)
      new Buffer(res.body).equals(GIF).must.be.true()
    })

    describe("given application/*", function() {
      it("must set body when Content-Type is application/octet-stream",
        function*() {
        var fetch = FetchParse(Fetch, {"application/*": true})
        var res = fetch("/")
        var headers = {"Content-Type": "application/octet-stream"}
        this.requests[0].respond(200, headers, GIF.toString("binary"))
        res = yield res
        res.body.must.be.an.instanceof(ArrayBuffer)
        new Buffer(res.body).equals(GIF).must.be.true()
      })
    })
  })
})


index.js
var MediaType = require("medium-type")
var JSONS = ["application/json", "*/*+json"].map(MediaType.parse)
var XMLS = ["application/xml", "text/xml", "*/*+xml"].map(MediaType.parse)
var WILDCARD_PARSER = [[new MediaType("*/*"), null]]
var PARSER_TYPE_ERR = "Parser not a function or true for default: "

exports = module.exports = function(fetch, types) {
  types = types == null ? WILDCARD_PARSER : parseParsers(types)
  return assign(exports.fetch.bind(null, fetch, types), fetch)
}

exports.fetch = function(fetch, types, url, opts) {
  return fetch(url, opts).then(exports.parse.bind(null, types))
}

exports.parse = function(types, res) {
  if (!hasContent(res)) return res
  var type = parseType(res.headers.get("content-type"))
  if (type == null) return res

  var parse = findParser(types, type)
  return parse ? parse(res).then(exports.set.bind(null, res)) : res
}

exports.arrayBuffer = function(res) {
  return res.arrayBuffer()
}

exports.text = function(res) {
  return res.text()
}

exports.json = function(res) {
  /* eslint consistent-return: 0 */
  return res.text().then(function(body) {
    if (body !== "") try { return JSON.parse(body) }
    catch (ex) { exports.set(res, body); throw errorify(res, ex) }
    else return undefined
  })
}

exports.set = function(res, body) {
  return Object.defineProperty(res, "body", {
    value: body, configurable: true, writable: true, enumerable: true
  })
}

function parseParsers(types) {
  return reduce(function(pairs, parser, type) {
    var types = expandType(type)

    if (parser === true)
      return pairs.concat(types.map(function(type) { return [type, null] }))
    else if (typeof parser == "function")
      return pairs.concat(types.map(function(type) { return [type, parser] }))
    else
      throw new TypeError(PARSER_TYPE_ERR + parser)
  }, [], types)
}

function expandType(type) {
  switch (type) {
    case "json": return JSONS
    case "xml": return XMLS
    default: return [new MediaType(type)]
  }
}

function hasContent(res) {
  if (res.bodyUsed) return false
  if (res.status === 204 || res.status === 304) return false
  var contentType = res.headers.get("content-type")
  return contentType != null && contentType !== ""
}

function parseType(type) {
  try { return new MediaType(type) }
  catch (ex) {
    if (ex instanceof SyntaxError || ex instanceof TypeError) return null
    else throw ex
  }
}

function findParser(types, type) {
  for (var i = 0; i < types.length; ++i)
    if (type.match(types[i][0])) return types[i][1] || getDefaultParser(type)

  return null
}

function getDefaultParser(type) {
  if (type.type === "text") return exports.text
  if (JSONS.some(type.match, type)) return exports.json
  if (XMLS.some(type.match, type)) return exports.text
  return exports.arrayBuffer
}

function errorify(res, err) {
  return Object.defineProperty(err, "response", {
    value: res, configurable: true, writable: true
  })
}

function reduce(fn, left, obj) {
  for (var key in obj) left = fn(left, obj[key], key)
  return left
}

function assign(a, b) { for (var k in b) a[k] = b[k]; return a }
